src/Installer.php

```diff
@@ -160,17 +177,59 @@ private function installPuli()
 
         $httpClient = new HttpClient($this->disableTls, $this->cafile);
 
+        $versionUrl = sprintf(
+            static::VERSION_API_URL,
+            $this->disableTls ? 'http' : 'https'
+        );
+
+        if (!$this->quiet) {
+            $this->info('Downloading...');
+        }
+
+        $versions = array();
+        for ($retries = 3; $retries > 0; --$retries) {
+            try {
+                $versions = $this->downloadVersions($httpClient, $versionUrl);
+                break;
+            } catch (RuntimeException $e) {
+                $this->error($e->getMessage());
+            }
+        }
+
+        if (0 === $retries || empty($versions)) {
+            $this->error('The download failed repeatedly, aborting.');
+
+            return 1;
+        }
+
+        $versionParser = new VersionParser($versions);
+        if (!empty($this->version)) {
+            if (!in_array($this->version, $versions, true)) {
+                $this->error(sprintf(
+                    'Could not find version: %s.',
+                    $this->version
+                ));
+
+                return 1;
+            }
+        } elseif ('stable' === $this->stability) {
+            $this->version = $versionParser->getMostRecentStable();
+            if (false === $this->version) {
+                $this->error('Could not find a stable version.');
+
+                return 1;
+            }
+        } else {
+            $this->version = $versionParser->getMostRecentAll();
+        }
+
         $url = sprintf(
-            '%s://github.com/puli/cli/releases/download/%s/puli.phar',
+            static::PHAR_DOWNLOAD_URL,
             $this->disableTls ? 'http' : 'https',
             $this->version
         );
 
         for ($retries = 3; $retries > 0; --$retries) {
-            if (!$this->quiet) {
-                $this->info('Downloading...');
```

Comment:
I think this block was inside the retry-block on purpose. Otherwise, if one try fails, the user will know about the error but not know that the installer is trying again. Hence I think this should be kept here and also added to the retry-loop above.